File: test-pdf-searchable.py
"""
Test both PDF export approaches:
1. Screenshot version (page.pdf) - works
2. Searchable version (exportPdfSearchable) - needs debugging
"""
from playwright.sync_api import sync_playwright
import time, os, json

URL = 'http://localhost:8765/index.html'
OUT_SCREENSHOT = '/home/saltedfish/project/cv/example/左秉鸿-screenshot-test.pdf'
OUT_SEARCHABLE = '/home/saltedfish/project/cv/example/左秉鸿-searchable-test.pdf'

def main():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True, args=['--no-sandbox'])
        page = browser.new_page(viewport={'width': 1280, 'height': 900})

        # Set download path
        context = browser.contexts[0]
        downloadPromise = [None]
        def on_download(download):
            downloadPromise[0] = download
        context.on('download', on_download)

        print('Loading page...')
        page.goto(URL, wait_until='domcontentloaded', timeout=60000)
        page.wait_for_selector('.resume-document[data-js-ready="true"]', timeout=15000)
        time.sleep(3)

        page_count = page.evaluate("document.querySelectorAll('.resume-page').length")
        print(f'Pages visible: {page_count}')

        stats = page.evaluate("""
            () => Array.from(document.querySelectorAll('.resume-page')).map((el, i) => {
                const c = el.querySelector('.resume-page-content');
                return {
                    i,
                    offsetH: el.offsetHeight,
                    contentScrollH: c ? c.scrollHeight : 0,
                    contentClientH: c ? c.clientHeight : 0,
                    banner: el.querySelector('.resume-page-banner span')?.textContent || 'N/A'
                };
            })
        """)
        print('Page stats:', json.dumps(stats, indent=2))

        # --- Test 1: Screenshot ---
        print('\n=== Test 1: Screenshot (page.pdf) ===')
        page.evaluate("document.getElementById('resumePages').style.gap = '0'")
        page.pdf(path=OUT_SCREENSHOT, format='A4',
                margin={'top': '8mm', 'bottom': '8mm', 'left': '8mm', 'right': '8mm'},
                print_background=True)
        sz1 = os.path.getsize(OUT_SCREENSHOT)
        print(f'Screenshot PDF: {sz1} bytes -> {OUT_SCREENSHOT}')

        # --- Test 2: Searchable ---
        print('\n=== Test 2: Searchable (exportPdfSearchable) ===')

        # Track console output
        console_logs = []
        page.on('console', lambda msg: console_logs.append(f'{msg.type}: {msg.text}'))

        print('Calling exportPdfSearchable()...')
        page.evaluate('exportPdfSearchable()')

        done = False
        for i in range(40):
            time.sleep(0.5)
            err = page.evaluate("window.__pdfError")
            if err:
                print(f'JS error: {err}')
                done = True
                break
            blob = page.evaluate("window.__pdfBlob")
            if blob:
                print(f'Blob captured: size={blob.get("size")}')
                data = page.evaluate("async () => Array.from(new Uint8Array(await window.__pdfBlob.arrayBuffer()))")
                with open(OUT_SEARCHABLE, 'wb') as f:
                    f.write(bytes(data))
                sz2 = os.path.getsize(OUT_SEARCHABLE)
                print(f'PDF saved: {sz2} bytes -> {OUT_SEARCHABLE}')
                done = True
                break
            # Check if download was triggered
            if downloadPromise[0]:
                dp = downloadPromise[0]
                dp.save_as(OUT_SEARCHABLE)
                sz2 = os.path.getsize(OUT_SEARCHABLE)
                print(f'Downloaded as: {sz2} bytes -> {OUT_SEARCHABLE}')
                done = True
                break
            if i % 5 == 4:
                print(f'  still waiting... {i+1}/40')

        if not done:
            print('No PDF blob or download after 20s')
            state = page.evaluate("() => ({ html2pdfDefined: typeof html2pdf !== 'undefined', pdfError: window.__pdfError })")
            print('State:', state)
            print('Console logs:', [l for l in console_logs if not l.startswith('log')])

        browser.close()
        if not done:
            exit(1)

if __name__ == '__main__':
    main()

File: test-pdf-searchable.mjs
import puppeteer from 'puppeteer';
import { writeFileSync } from 'fs';

const url = 'http://localhost:8765/index.html';
const outPath = '/home/saltedfish/project/cv/example/左秉鸿-searchable-test.pdf';

async function main() {
  const browser = await puppeteer.launch({ headless: true, args: ['--no-sandbox'] });
  const page = await browser.newPage();

  await page.setViewport({ width: 1280, height: 900 });

  console.log('Loading page...');
  await page.goto(url, { waitUntil: 'networkidle0', timeout: 30000 });

  await page.waitForSelector('.resume-document[data-js-ready="true"]', { timeout: 15000 });
  await new Promise(r => setTimeout(r, 2000));

  const pageCount = await page.$$eval('.resume-page', els => els.length);
  console.log(`Pages visible: ${pageCount}`);

  const pageStats = await page.$$eval('.resume-page', els => els.map(el => ({
    height: el.scrollHeight,
    clientHeight: el.clientHeight,
    contentScrollH: el.querySelector('.resume-page-content')?.scrollHeight || 0,
    contentClientH: el.querySelector('.resume-page-content')?.clientHeight || 0,
    banner: el.querySelector('.resume-page-banner span')?.textContent || 'N/A',
    children: el.querySelector('.resume-page-content')?.children.length || 0
  })));
  console.log('Page stats:', JSON.stringify(pageStats, null, 2));

  // Capture all console output from the page
  const logs = [];
  page.on('console', msg => logs.push(msg.type() + ': ' + msg.text()));

  console.log('Calling exportPdfSearchable()...');

  // Intercept the download and capture the PDF binary
  const client = await page.createCDPSession();
  await client.send('Page.setDownloadBehavior', {
    behavior: 'allow',
    downloadPath: '/tmp'
  });

  // Set up a response interceptor for the PDF save
  let pdfSaved = false;
  let pdfBuffer = null;

  // Override Blob to capture the save data
  await page.evaluateOnNewDocument(() => {
    const originalSave = HTMLAnchorElement.prototype.click;
    window.__pdfBlobs = [];
    window.__interceptedBlob = null;

    // Listen for Blob URL creation
    const oldCreateObjectURL = URL.createObjectURL.bind(URL);
    URL.createObjectURL = function(blob) {
      const result = oldCreateObjectURL(blob);
      if (blob.type === 'application/pdf') {
        window.__interceptedBlob = blob;
      }
      return result;
    };
  });

  // Call the export function
  await page.evaluate('exportPdfSearchable()');

  // Wait for the PDF to be generated (give it up to 15 seconds)
  await new Promise(r => setTimeout(r, 15000));

  // Try to capture the blob
  const blobInfo = await page.evaluate(() => {
    if (window.__interceptedBlob) {
      return {
        size: window.__interceptedBlob.size,
        type: window.__interceptedBlob.type
      };
    }
    return null;
  });

  if (blobInfo) {
    console.log('Intercepted blob:', blobInfo);
    // Read blob as array buffer
    const arrayBuffer = await page.evaluate(async (blob) => {
      return await blob.arrayBuffer();
    }, await page.evaluate(() => window.__interceptedBlob));

    const buffer = Buffer.from(arrayBuffer);
    writeFileSync(outPath, buffer);
    console.log(`PDF saved: ${buffer.length} bytes to ${outPath}`);
    pdfSaved = true;
  } else {
    // Fallback: check if file was downloaded to /tmp
    console.log('Checking /tmp for downloaded PDF...');
    try {
      const files = await page.evaluate(() => {
        // Try to find PDF files
        return window.__pdfBlobs;
      });
      console.log('Captured blobs:', files);
    } catch (e) {
      console.log('Could not capture blob:', e.message);
    }

    // Check page logs
    console.log('\nPage console logs:');
    logs.forEach(l => console.log(' ', l));

    // Check if PDF was saved via download
    const downloadCheck = await page.evaluate(() => {
      // Try to read the last downloaded file via FileReader API
      const links = document.querySelectorAll('a[download]');
      return Array.from(links).map(a => ({ href: a.href, download: a.download }));
    });
    console.log('Download links found:', downloadCheck);
  }

  // Print any console logs
  if (logs.length) {
    console.log('\nConsole output:', logs);
  }

  await browser.close();

  if (!pdfSaved) {
    process.exit(1);
  }
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});